drivers/usb/device/usb_dc_nrfx.c

```diff
@@ -0,0 +1,1580 @@
+		memset(ep_ctx, 0, sizeof(*ep_ctx));
+	}
+}
+
+
+/**
+ * @brief Enable or disable USBD power interrupt.
+ *
+ * @param enable New interrupt enable state.
+ */
+static int usbd_power_int_enable(bool enable)
+{
+	struct device *dev = device_get_binding(CONFIG_USBD_NRF5_NAME);
+
+	if (!dev) {
+		SYS_LOG_ERR("Could not get USBD power device binding.");
+		return -ENODEV;
+	}
+
+	nrf5_power_usb_power_int_enable(dev, enable);
+
+	return 0;
+}
+
+
+static void usbd_event_transfer_ctrl(nrfx_usbd_evt_t const * const p_event)
+{
+	struct nrf5_usbd_ep_ctx *ep_ctx =
+		endpoint_ctx(p_event->data.eptransfer.ep);
+	struct nrf5_usbd_ctx *ctx = get_usbd_ctx();
+
+	if (NRF_USBD_EPIN_CHECK(p_event->data.eptransfer.ep)) {
+		if (p_event->data.eptransfer.status == NRF_USBD_EP_OK) {
+			struct usbd_ep_event *ev = usbd_evt_alloc();
+
+			ev->ep = ep_ctx;
+			ev->evt = EP_EVT_WRITE_COMPLETE;
+
+			k_sem_give(&ctx->dma_in_use);
+			SYS_LOG_DBG("ctrl write complete");
+			usbd_evt_put(ev);
+			usbd_work_schedule();
+		}
+	} else {
+		if (p_event->data.eptransfer.status == NRF_USBD_EP_WAITING) {
+			struct usbd_ep_event *ev = usbd_evt_alloc();
+
+			SYS_LOG_DBG("ctrl read request");
+
+			ep_ctx->read_pending = true;
+			ev->ep = ep_ctx;
+			ev->evt = EP_EVT_RECV_REQ;
+			usbd_evt_put(ev);
+			usbd_work_schedule();
+		} else if (p_event->data.eptransfer.status == NRF_USBD_EP_OK) {
+			struct usbd_ep_event *ev = usbd_evt_alloc();
+			nrfx_err_t err_code;
+
+			ev->ep = ep_ctx;
+			ev->evt = EP_EVT_RECV_COMPLETE;
+
+			err_code = nrfx_usbd_ep_status_get(
+				p_event->data.eptransfer.ep, &ep_ctx->buf.len);
+
+			if ((err_code != NRFX_SUCCESS) &&
+				(err_code != (nrfx_err_t)NRF_USBD_EP_OK)) {
+				SYS_LOG_ERR("_ep_status_get failed! Code: %d.",
+					    err_code);
+				__ASSERT_NO_MSG(0);
+			}
+			SYS_LOG_DBG("ctrl read done: %d", ep_ctx->buf.len);
+
+			k_sem_give(&ctx->dma_in_use);
+			usbd_evt_put(ev);
+			usbd_work_schedule();
+		}
+	}
+}
+
+
+static void usbd_event_transfer_data(nrfx_usbd_evt_t const * const p_event)
+{
+	struct nrf5_usbd_ctx *ctx = get_usbd_ctx();
+	struct nrf5_usbd_ep_ctx *ep_ctx =
+		endpoint_ctx(p_event->data.eptransfer.ep);
+
+	if (NRF_USBD_EPIN_CHECK(p_event->data.eptransfer.ep)) {
+		if (p_event->data.eptransfer.status == NRF_USBD_EP_OK) {
+			struct usbd_ep_event *ev = usbd_evt_alloc();
+
+			ev->ep = ep_ctx;
+			ev->evt = EP_EVT_WRITE_COMPLETE;
+
+			SYS_LOG_DBG("write complete, ep %d",
+				    (u32_t)p_event->data.eptransfer.ep);
+			k_sem_give(&ctx->dma_in_use);
+			usbd_evt_put(ev);
+			usbd_work_schedule();
+		}
+	} else {
+		if (p_event->data.eptransfer.status == NRF_USBD_EP_WAITING) {
+			struct usbd_ep_event *ev = usbd_evt_alloc();
+
+			SYS_LOG_DBG("read request, ep %d",
+				    (u32_t)p_event->data.eptransfer.ep);
+
+			ev->ep = ep_ctx;
+			ev->evt = EP_EVT_RECV_REQ;
+			usbd_evt_put(ev);
+			ep_ctx->read_pending = true;
+			usbd_work_schedule();
+		} else if (p_event->data.eptransfer.status == NRF_USBD_EP_OK) {
```

Comment:
And what about the other statuses: `NRF_USBD_EP_OVERLOAD` and `NRF_USBD_EP_ABORTED`. Shouldn't these also be handled (at least regarding `ctx->dma_in_use`)?